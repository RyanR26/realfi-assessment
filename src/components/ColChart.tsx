import React, { useMemo } from 'react';
import objArrayValueToArray from '../utils/objArrayValueToArray'

interface IProps {
  dataPoints: IDataPoint[],
  config?: {
    title?: string,
    xLabel? :string,
    tooltipLabel?: string,
    height?: string,
    width?: string
  }
}

interface IDataPoint { 
  x: string | number, 
  y: number,
  label?: string
}

export default function ColChart(props: IProps) {

  const { 
    dataPoints,
    config 
  } = props;
    
  // Sort age groups correctly - Had to use type 'any' with sort method to satisfy TS compiler 
  const dataPointsSorted = useMemo(() => dataPoints.sort((a:any, b:any) => (a.x || '').localeCompare(b.x || '')), [dataPoints]);
  const yValues = useMemo<number[]>(() => objArrayValueToArray(dataPoints, 'y'), [dataPoints]);
  const maxYValue = useMemo(() => Math.max(...yValues), [yValues]);
  const height = config?.height || '500px';
  const width = config?.width || '100%';

  return (
    <div className='col-chart'>
      { config?.title && <div className='text-md text-center pb-4 underline'>{ config.title }</div> }
      <div className='cols-container flex justify-evenly items-start border-2 rounded-box border-primary p-4' style={{ height, width }}>
      {
        dataPointsSorted.map((dataPoint: IDataPoint) => (
          <div key={dataPoint.x} className='h-full flex relative w-1/3'>
            <div 
            className='mt-auto mx-auto mb-0 w-1/3 bg-secondary flex justify-center text-black rounded-lg pt-4 tooltip'
            data-tip={`${dataPoint.y} ${config?.tooltipLabel}`} 
            style={{ 
              height: `${dataPoint.y/maxYValue * 100}%`,
              transition: 'all 500ms ease'
            }}
            >
              <span className='text-xs'>{dataPoint.y}</span>
            </div>
            <div className='absolute bottom-[-50px] ml-[50%] translate-x-[-50%]'>{dataPoint.x}</div>
          </div> 
        ))
      }
      </div>
      { config?.xLabel && <div className='text-xs text-center border-t border-dashed border-primary pt-[10px] mt-[40px]'>{config.xLabel}</div> }
    </div>
  )
};