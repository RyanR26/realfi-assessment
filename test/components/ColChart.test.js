import renderer from 'react-test-renderer'
import ColChart from '../../src/components/ColChart'

it('Should render the ColChart HTML correctly against the snapshot', () => {

  const colChartDataPoints = [
    {
        "y": 2,
        "x": "21-30"
    },
    {
        "y": 1,
        "x": "31-40"
    },
    {
        "y": 2,
        "x": "41-50"
    }
  ];

  const config = {
    "title": "Users by age group",
    "xLabel": "Ages in years",
    "tooltipLabel": "User(s)",
    "width": "500px",
    "height": "300px"
  };

  const component = renderer.create(<ColChart dataPoints={colChartDataPoints} config={config}/ >);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  {/* // manually trigger the callback
  renderer.act(() => {
    tree.props.onMouseEnter();
  });
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger the callback
  renderer.act(() => {
    tree.props.onMouseLeave();
  });
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot(); */}
});