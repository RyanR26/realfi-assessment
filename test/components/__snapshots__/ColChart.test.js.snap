// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the ColChart HTML correctly against the snapshot 1`] = `
<div
  className="col-chart"
>
  <div
    className="text-md text-center pb-4 underline"
  >
    Users by age group
  </div>
  <div
    className="cols-container flex justify-evenly items-start border-2 rounded-box border-primary p-4"
    style={
      {
        "height": "300px",
        "width": "500px",
      }
    }
  >
    <div
      className="h-full flex relative w-1/3"
    >
      <div
        className="mt-auto mx-auto mb-0 w-1/3 bg-secondary flex justify-center text-black rounded-lg pt-4 tooltip"
        data-tip="2 User(s)"
        style={
          {
            "height": "100%",
            "transition": "all 500ms ease",
          }
        }
      >
        <span
          className="text-xs"
        >
          2
        </span>
      </div>
      <div
        className="absolute bottom-[-50px] ml-[50%] translate-x-[-50%]"
      >
        21-30
      </div>
    </div>
    <div
      className="h-full flex relative w-1/3"
    >
      <div
        className="mt-auto mx-auto mb-0 w-1/3 bg-secondary flex justify-center text-black rounded-lg pt-4 tooltip"
        data-tip="1 User(s)"
        style={
          {
            "height": "50%",
            "transition": "all 500ms ease",
          }
        }
      >
        <span
          className="text-xs"
        >
          1
        </span>
      </div>
      <div
        className="absolute bottom-[-50px] ml-[50%] translate-x-[-50%]"
      >
        31-40
      </div>
    </div>
    <div
      className="h-full flex relative w-1/3"
    >
      <div
        className="mt-auto mx-auto mb-0 w-1/3 bg-secondary flex justify-center text-black rounded-lg pt-4 tooltip"
        data-tip="2 User(s)"
        style={
          {
            "height": "100%",
            "transition": "all 500ms ease",
          }
        }
      >
        <span
          className="text-xs"
        >
          2
        </span>
      </div>
      <div
        className="absolute bottom-[-50px] ml-[50%] translate-x-[-50%]"
      >
        41-50
      </div>
    </div>
  </div>
  <div
    className="text-xs text-center border-t border-dashed border-primary pt-[10px] mt-[40px]"
  >
    Ages in years
  </div>
</div>
`;
